### PR TITLE
ENH,WIP: add FactorialPartitioner to allow hierarchical cv

### PR DESCRIPTION
**Pull request by mvdoc**
With code from Yarik who created it in an afternoon of madness. Docstring still WIP, submitted to track it. Unittest is simply a check against the old chain of nfoldpartitioner and sifter.

### CODE REVIEW

File: mvpa2/generators/partition.py [hunk 1]
```diff
@@ -491,6 +493,54 @@ def _get_partition_specs(self, uniqueattrs):
                     if ind < self.count]
 
 
+class FactorialPartitioner(Partitioner):
+    """
```
Comment (yarikoptic) [inline mvpa2/generators/partition.py]: Provide one line summary description with empty line under,e.g.

```
"""Partitioner for two-level factorial designs

Given ...
"""
```

File: mvpa2/generators/partition.py [hunk 2]
```diff
@@ -491,6 +493,54 @@ def _get_partition_specs(self, uniqueattrs):
...
+    Given another partitioner on a dataset with two hierarchically organized
+    attributes, generates the combinations of all possible folds of the
+    superordinate attribute with the balanced combinations of the subordinate
+    attribute. For example, splitting familiar and unfamiliar faces while
+    balancing identities in the training/test set.
+
+    ADD EXAMPLE HERE
```
Comment (yarikoptic) [inline mvpa2/generators/partition.py]: YES -- would be super useful.  Verify that it is "runnable" by running doctests

nosetests -s -v --with-doctest --doctest-tests mvpa2/generators/partition.py
```

File: mvpa2/generators/partition.py [hunk 3]
```diff
@@ -491,6 +493,54 @@ def _get_partition_specs(self, uniqueattrs):
...
+            def copy(self, deep=False):
+                assert(not deep)  # dunno how
+                ds = FakeDataset(self)
+                ds.sa = ds.sa.copy(deep=False)
+                return ds
+
+            def __getitem__(self, item):
+                return FakeDataset(self[item])
+
+        #fakeds = FakeDataset(ds)
```
Comment (yarikoptic) [inline mvpa2/generators/partition.py]: kill FakeDataset above since not used, and provide description under for why selecting a single feature

File: mvpa2/tests/test_generators.py [hunk 4]
```diff
@@ -396,3 +397,46 @@ def is_sorted(x):
...
+    # now the new implementation
+    factpart = FactorialPartitioner(
+        NFoldPartitioner(attr='subord'),
+        attr='superord'
+    )
+
+    partitions_npart = [p.sa.partitions for p in npart.generate(ds)]
+    partitions_factpart = [p.sa.partitions for p in factpart.generate(ds)]
+
+    assert_array_equal(np.sort(partitions_npart), np.sort(partitions_factpart))
```
Comment (yarikoptic) [inline mvpa2/tests/test_generators.py]: we should test on few more cases, ideally corner cases.  E.g. What would happen if there is a single superordinate class?

would it blow or not, or issue a warning, if in one superord class we don't have all the subord categories? ;)

Also for completely dummy verification -- e.g. for 2x2 design verify that actual partitionings are as expected. (so we have a test which would not rely on correctness of NFold+Sifter construct operation)

File: mvpa2/generators/partition.py [hunk 5]
```diff
@@ -491,6 +493,54 @@ def _get_partition_specs(self, uniqueattrs):
...
+    Given another partitioner on a dataset with two hierarchically organized
+    attributes, generates the combinations of all possible folds of the
+    superordinate attribute with the balanced combinations of the subordinate
+    attribute. For example, splitting familiar and unfamiliar faces while
+    balancing identities in the training/test set.
+
+    ADD EXAMPLE HERE
+    """
+    def __init__(self, partitioner, **kwargs):
+        Partitioner.__init__(self, **kwargs)
```
Comment (yarikoptic) [inline mvpa2/generators/partition.py]: to be safe, use `super` construct (even if other Partitioners use old fashion way like this, don't remember) -- we do have multiple inheritence around the hierarchy

File: mvpa2/tests/test_generators.py (hunk 4 above)
Comment (mvdoc) [inline mvpa2/tests/test_generators.py]: Added test 1 and 3. For the second case, I tried and it doesn't work of course. My take is that it should blow when one superord class doesn't have all subord. What do you think?

File: mvpa2/generators/partition.py [hunk 6]
```diff
@@ -491,6 +493,45 @@ def _get_partition_specs(self, uniqueattrs):
                     if ind < self.count]
 
 
+class FactorialPartitioner(Partitioner):
+    """Partitioner for two-level factorial designs
+
+    Given another partitioner on a dataset with two hierarchically organized
+    attributes, generates the combinations of all possible folds of the
+    superordinate attribute with the balanced combinations of the subordinate
+    attribute. For example, splitting familiar and unfamiliar faces while
```
Comment (yarikoptic) [inline mvpa2/generators/partition.py]: A bit more explicit, that splitting to cross validate across identities, because splitting faces is not clear... And that is under example. Use

```
Example
-----------
```

File: mvpa2/tests/test_generators.py (hunk 4 above)
Comment (yarikoptic) [inline mvpa2/tests/test_generators.py]: I actually wonder... Probably it shouldn't blow if superords aren't balancedd in subords, since it might be a legit case and people might decide to use it with Balancer to mitigate

Comment (mvdoc) [inline mvpa2/tests/test_generators.py]: so just a warning?

Comment (yarikoptic) [inline mvpa2/tests/test_generators.py]: At most

**Comment by mvdoc** [comment]
@yarikoptic let me know if this is ok

File: mvpa2/tests/test_generators.py [hunk 7]
```diff
@@ -396,3 +397,79 @@ def is_sorted(x):
...
+
+    # now let's check it behaves correctly if we have only one superord class
+    nfold = NFoldPartitioner(attr='subord')
+    partitions_nfold = [p.sa.partitions for p in nfold.generate(ds_1super)]
+    partitions_factpart = [p.sa.partitions for p in factpart.generate(ds_1super)]
+    assert_array_equal(np.sort(partitions_nfold), np.sort(partitions_factpart))
+
+    # smoke test for unbalanced subord classes
+    partitions_factpart = [p.sa.partitions
+                           for p in factpart.generate(ds_unbalanced)]
```
Comment (yarikoptic) [inline mvpa2/tests/test_generators.py]: and what happens -- I guess warning is issued...  could you check if our assert_warnings works for this one? ;)
For that please do not use our ad-hoc mvpa2.base warning but `warnings.warn(msg, category=warnings.RuntimeWarning)` function, and here is an example from another place how to assert
```
from mvpa2.testing import assert_warnings
...
        with assert_warnings(
                [(DeprecationWarning,
                  "allowedtype option was deprecated in favor of constraints. "
                  "Adjust your code, provided value 'str' was ignored")]):
            DO_SOMETHIONG_WHICH_SPITS_WARNING
```

and what is the output? ;-) better to "hardcode" expectation here

Then I will become happy ;-)

**Comment by mvdoc** [comment]
@yarikoptic better now? ;-)
